src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -0,0 +1,213 @@
+        return new InfinispanJF( file, lockOwner, callbacks, doOutput, opLock, this );
+    }
+
+    @Override
+    public LocalLockManager getLocalLockManager()
+    {
+        return null;
+    }
+
+    FileBlock getNextBlock( final FileBlock prevBlock, final FileMeta metadata) throws IOException
+    {
+        synchronized ( InfinispanJFS.this )
+        {
+            // TODO: make sure next will exist
+            UUID next = prevBlock.getNextBlockID();
+            FileBlock nextBlock = null;
+            while ( nextBlock == null )
+            {
+                nextBlock = blockCache.get( next );
+                if (nextBlock == null)
+                {
+                    try
+                    {
+                        lockManager.reentrantSynchronous( metadata.getFilePath(), ( opLock ) -> {
+                            // setup a cache listener for this UUID, and wait in a timed loop for it to return
+                            ClusterListener clusterListener = new ClusterListener( next, opLock );
+                            clusterListener.listenToCache( blockCache );
+                            opLock.await( WAIT_TIMEOUT );
+
+                            return null;
+                        } );
+                    }
+                    catch ( final InterruptedException e )
+                    {
+                        return null;
+                    }
+                }
+
+            }
+
+            return nextBlock;
+        }
+    }
+
+    void pushNextBlock( final FileBlock prevBlock, final FileBlock nextBlock, final FileMeta metadata )
+                    throws IOException
+    {
+        try
+        {
+            lockManager.reentrantSynchronous( metadata.getFilePath(), ( opLock ) -> {
+                blockCache.put( prevBlock.getBlockID(), prevBlock );
+                blockCache.put( nextBlock.getBlockID(), nextBlock );
+                metadataCache.put( metadata.getFilePath(), metadata );
+
+                return null;
+            } );
+        }
+        catch ( InterruptedException e )
+        {
+            throw new IOException( "Thread interrupted while adding next block to file: " + metadata.getFilePath(), e );
+        }
+    }
+
+    void updateBlock( final FileBlock block ) throws IOException
+    {
+        blockCache.put( block.getBlockID(), block );
+    }
+
+    void close( FileMeta metadata ) throws IOException
+    {
+        try
+        {
+            lockManager.reentrantSynchronous( metadata.getFilePath(), ( opLock ) -> {
+
+                // called from InfinispanJF.reallyClose()? I think.
+                // if the local lock owner is empty / lock count == 0 for this file, remove it from the node-level locks too.
+                // if the local lock owner's count is NOT 0, then we need to make sure our node's lock-level matches the one in
+                // the local lock manager, and update if necessary
+                LocalLockOwner owner = metadata.getLockOwner();
```

Comment:
At this point, will LocalLockOwner.unlock() have been called? If not, then I don't think we can track dominant lock level changes...we might need to handle this in FileTree after calling LocalLockOwner.unlock(). See JIRA comments.